libs/gl-client/src/pairing/attestation_device.rs

```diff
@@ -148,4 +150,119 @@ impl Client<Connected> {
             .await?
             .into_inner())
     }
+
+    pub fn verify_pairing_data(data: GetPairingDataResponse) -> Result<()> {
+        let mut crs = std::io::Cursor::new(&data.csr);
+        let pem = Pem::read_from(&mut crs).map_err(into_verify_pairing_data_error)?;
+        let csr = Csr::from_pem(&pem).map_err(into_verify_pairing_data_error)?;
+        let sub_pk_der = csr
+            .public_key()
+            .to_der()
+            .map_err(into_verify_pairing_data_error)?;
+        let sub_pk_info: SubjectPublicKeyInfo =
+            picky_asn1_der::from_bytes(&sub_pk_der).map_err(into_verify_pairing_data_error)?;
+
+        if let PublicKey::Ec(bs) = sub_pk_info.subject_public_key {
+            let pk = hex::encode(bs.0.payload_view());
+
+            if pk == data.session_id
+                && Self::restriction_contains_pubkey_exactly_once(&data.restrs, &data.session_id)
+            {
+                Ok(())
+            } else {
+                Err(Error::VerifyPairingDataError(format!(
+                    "public key {} does not match pk {}",
+                    data.session_id, pk
+                )))
+            }
+        } else {
+            Err(Error::VerifyPairingDataError(format!(
+                "public key is not ecdsa"
+            )))
+        }
+    }
+
+    /// Checks that a restriction string only contains a pubkey field exactly
+    /// once that is not preceded or followed by a '|' to ensure that it is
+    /// not part of an alternative but a restriction by itself.
+    fn restriction_contains_pubkey_exactly_once(s: &str, pubkey: &str) -> bool {
+        let search_field = format!("pubkey={}", pubkey);
+        match s.find(&search_field) {
+            Some(index) => {
+                // Check if 'pubkey=<pubkey>' is not preceded by '|'
+                if index > 0 && s.chars().nth(index - 1) == Some('|') {
+                    return false;
+                }
+
+                // Check if 'pubkey=<pubkey>' is not followed by '|'
+                let end_index = index + search_field.len();
+                if end_index < s.len() && s.chars().nth(end_index) == Some('|') {
+                    return false;
+                }
+
+                // Check if 'pubkey=<pubkey>' appears exactly once
+                s.matches(&search_field).count() == 1
+            }
+            None => false,
+        }
+    }
```

Comment:
Shouldn't we be checking for a standalone `pubkey=03....` restriction? That'd mean that there cannot be any qualifiers (alternatives) on that restriction, and any rune including this restriction has to adhere to that pubkey. Attempting to parse a logical expression, and testing it doesn't somehow negate the clause we're interested in is hard (NP-hard to be precise, even if parsing works perfectly).